src/natcap/invest/hra.py

```diff
@@ -1,3253 +1,2424 @@
+                raster.
+
+        Returns:
+            out_array (numpy.array): An unsigned byte mask with pixel values of
+            0 (on nodata pixels) or 1 (on non-nodata pixels).
+        """
+        return (raster_values != nodata).astype(numpy.uint8)
+
+    pygeoprocessing.raster_calculator(
+        [(source_raster_path, 1)], _rewrite, target_raster_path,
+        gdal.GDT_Byte, 0)
+
+
+def _polygonize(source_raster_path, mask_raster_path,
+                target_polygonized_vector, field_name, layer_name):
+    """Polygonize a raster.
+
+    Args:
+        source_raster_path (string): The source raster to polygonize.  This
+            raster must be an integer raster.
+        mask_raster_path (string): The path to a mask raster, where pixel
+            values are 1 where polygons should be created, 0 otherwise.
+        target_polygonized_vector (string): The path to a vector into which the
+            new polygons will be inserted.  A new GeoPackage will be created at
+            this location.
+        field_name (string): The name of a field into which the polygonized
+            region's numerical value should be recorded.  A new field with this
+            name will be created in ``target_polygonized_vector``, and with an
+            Integer field type.
+        layer_name (string): The layer name to use in the target vector.
+
+    Returns:
+        ``None``
+    """
+    LOGGER.info(f'Polygonizing {source_raster_path} --> '
+                f'{target_polygonized_vector} using mask {mask_raster_path}')
+    raster = gdal.OpenEx(source_raster_path, gdal.OF_RASTER)
+    band = raster.GetRasterBand(1)
+
+    mask_raster = gdal.OpenEx(mask_raster_path, gdal.OF_RASTER)
+    mask_band = mask_raster.GetRasterBand(1)
+
+    raster_srs = osr.SpatialReference()
+    raster_srs.ImportFromWkt(raster.GetProjectionRef())
+
+    driver = gdal.GetDriverByName('GPKG')
+    vector = driver.Create(
+        target_polygonized_vector, 0, 0, 0, gdal.GDT_Unknown)
+    layer = vector.CreateLayer(layer_name, raster_srs, ogr.wkbPolygon)
+
+    # Create an integer field that contains values from the raster
+    field_defn = ogr.FieldDefn(str(field_name), ogr.OFTInteger)
+    field_defn.SetWidth(3)
+    field_defn.SetPrecision(0)
+    layer.CreateField(field_defn)
+
+    raster_srs = osr.SpatialReference()
+    raster_srs.ImportFromWkt(raster.GetProjectionRef())
+
+    layer.StartTransaction()
+    # 0 represents field index 0, into which pixel values will be written.
+    gdal.Polygonize(band, mask_band, layer, 0)
+    layer.CommitTransaction()
+
+
+def _create_summary_statistics_file(
+        aoi_raster_json_path, habitat_mask_raster_path, pairwise_raster_dicts,
+        target_summary_csv_path):
+    """Summarize pairwise habitat/stressor rasters by AOI subregions.
+
+    Functionally, this is a modified version of
+    ``pygeoprocessing.zonal_statistics`` that doesn't force re-rasterization of
+    subregions.
```

Comment:
Yeah, this functionality follows what's in the current HRA (which I'm not committed to), which attempts to avoid potentially many re-rasterizations in order to do `zonal_statistics`.  The only differentiation between this and `pygeoprocessing.zonal_statistics` is that this function attempts to support grouping features with the same name.

That being said, all of this rasterization/custom stats workaround is a pain and probably one of the harder things to maintain in this model and it doesn't have to be.  I'll rework this into `zonal_statistics` and get back to you.

Reply:
@davemfish after digging into this a bit more, the core issue is that `zonal_statistics` does not currently provide enough information to be able to build the `R_%LOW`, `R_%MED`, and `R_%HIGH` columns; they cannot be generally derived from the `sum` and `count` fields returned from zonal_stats.  I've put together a [gist going into a bit more detail about it here](https://gist.github.com/phargogh/aff32bb689220fb317b0a5196c7cf307).

So if you're OK with it, I'd prefer to leave this function as-is until I can propose a solution within `zonal_statistics` and we can ship it.